docs/proposals/enable-local-source-code-support.md

```diff
@@ -0,0 +1,231 @@
+
+This EP is well aware of all potential changes for the Build API in the future, such as:
+
+- [Remote Artifacts](https://github.com/shipwright-io/build/blob/master/docs/proposals/remote-artifacts.md)
+- [Build Inputs Overhaul](https://github.com/shipwright-io/build/pull/652)
+
+But to be pragmatic and avoid confusion, this EP proposes to do the following changes in the current state of the Build API.
+
+##### Modify the API
+
+- Introduce `spec.source.local`, which should host the desire usage of this feature. The value of `spec.source.local` requires to be an absolute path to the code in the user local machine. See an [example](https://github.com/qu1queee/build/blob/qu1queee/local_source/pkg/apis/build/v1alpha1/source.go#L14-L16).
+
+- Ensure `spec.source.url` is not longer a mandatory field. This was done thinking on assets only hosted in `git`, which no longer holds true. See an [example](https://github.com/qu1queee/build/blob/qu1queee/local_source/pkg/apis/build/v1alpha1/source.go#L19-L20).
+
+#### Modify the Runtime logic
+
+- Ensure that when `spec.source.local` is defined, we do not longer generate the Tekton Input PipelineResource. We do this today, to tell Tekton that we want to pull source from a git repository, which ends as a container that pulls it. See an [example](https://github.com/qu1queee/build/blob/qu1queee/local_source/pkg/reconciler/buildrun/resources/taskrun.go#L184-L185) on future changes.
+
+- For the bundle mechanism, which is explained in the next section. We need to **prepend** a new step in our Task step definition, which will pull our local source code from a registry. See an [example](https://github.com/qu1queee/build/blob/qu1queee/local_source/pkg/reconciler/buildrun/resources/taskrun.go#L171-L186). 
+
+  Important to notice, the image to pull will be a self-extracted image, therefore the `workingDir` container definition should be under `/workspace/source`, which is a well-known path in the Shipwright strategies, where source code is expected to be.
+
+_Note_: More changes code-wise might be needed, but that should be an implementation detail and does not belong in this EP description.
+_Note_: The above does not require a manual modification to any of the existing Shipwright strategies.
+
+
+### Part 2: Bundle Mechanism
+
+This is CLI specific. Any workflow using Shipwright that intends to provide support for local source code will require to re-use the Shipwright/CLI implementation or implement a standalone one.
+
+This proposed implementation can be one of many. In the future we might want to use a different storage service as needed.
+
+The Bundle mechanism has it´s origins from [Mink Bundles](https://github.com/mattmoor/mink/tree/master/pkg/bundles) and it´s based on the premise that if a user is pushing a container image to a container registry, then workflows could use the same approach to push the local source code, and pull it at a later stage during the building image mechanism.
```

Comment:
If we're going to rely on this, we should probably formalize and understand the format better ourselves. In particular, I have a lurking concern that depending on a container registry for this could lead to storing lots of data in a registry, and being charged for it. GCR doesn't garbage-collect automatically, for example.

I'm also not sure if mink bundles do anything smart to avoid limits on the number of layers a container image can have, or if it just keeps piling layers on top of a previous copy forever. It's something we should understand better at least.

Using mink bundles can also make it harder to track the source that went into a build, so we should recommend users not use it for things they might want to track down and audit/debug later. Git is a much better choice for that.

(These aren't reasons not to use mink, I just want to make sure we understand what we're getting into by adopting it as our solution for local source builds)

Reply:
I am curious if mink development is continuing now that Matt has stepped back. Are there plans to fork and/or contribute to it?

Reply:
@lance that's a very good question. I think if we proceed with using bundles (not necessarily all of `mink`) we should work with Matt to move it to some neutral location where it can be more of a community resource.

That could be within Shipwright, or perhaps Tekton, or perhaps some new entity we spin up.

Reply:
@ImJasonH for the excessive storage concern, we can always append a final step that cleans-up the bundle image. I´m wondering if this would be something you will opt-in. Users could decide if they want this or not.

I agree on understanding more the details on the layers. I try to highlight the need of not copying what the Bundles pkg in Mink have, but rather have our custom implementation. I think that as part of doing this, we will tackle this concern. 

I also modify this EP to not mention `Mink`, and only references to a `Bundles` package in that repository. The reason is that this is not at all about Mink, I think Mink have much more features, this EP only proposes to re-use the ideas of Bundle in Mink.

Reply:
> I also modify this EP to not mention Mink, and only have references to a Bundles package in that repository. The reason is that this is not at all about Mink, I think Mink have much more features, this EP only proposes to re-use the ideas of Bundle in Mink.

Thanks - I think keeping Mink out of the EP will eliminate some confusion, and also better allow for deviation from that approach if need be.

Reply:
If we take a dependency on the mink bundle implementation, I think we should fork that code into an internal package in shipwright-io/build, so we can own and iterate on the code instead of depending on @mattmoor's repo.

I'm also a bit uncomfortable automatically cleaning up old bundles by default, since this loses provenance information. I guess registries that aggressively garbage-collect (Quay, DockerHub, but notably _not_ GCR) could do the right thing for us, where `shp build <local source>` could optionally tag the bundle image to keep it from being GCed, or only reference it by digest where the image would get deleted by the registry after some time.

Reply:
@ImJasonH we dont have a dependency, we can have the code from scratch in our side. That would also provide us more control on defining and maintaining the base image layer for the bundle image, which I think is not even defined under the bundle pkg, but somewhere else.

Do you want to agree on the cleanup on this EP, or it is fine if we take this further during the development cycles of the EP? I can ensure the EP includes the concerns on excessive storage.